File: src/bridge/Bridge.cpp
#include "Bridge.hpp"
#include <nlohmann/json.hpp>
#include <sstream>

using json = nlohmann::json;

Bridge::Bridge(Calculator& calculator) : m_calc(calculator) {}

std::string Bridge::handleMessage(const std::string& action, const std::string& jsonArgs) {
    try {
        if (action == "calculate") {
            auto args = json::parse(jsonArgs);
            std::string expr = args[0].get<std::string>();
            std::string angleUnit = "degrees";
            if (args.size() > 1) angleUnit = args[1].get<std::string>();
            return handleCalculate(expr, angleUnit);
        } else if (action == "memory") {
            auto args = json::parse(jsonArgs);
            std::string op = args[0].get<std::string>();
            double val = (args.size() > 1) ? args[1].get<double>() : 0.0;
            return handleMemory(op, val);
        } else if (action == "angleUnit") {
            auto args = json::parse(jsonArgs);
            std::string unit = args[0].get<std::string>();
            return handleAngleUnit(unit);
        } else {
            json err = {{"type", "error"}, {"code", "UNKNOWN_ACTION"}, {"message", "Unknown action: " + action}};
            return err.dump();
        }
    } catch (const std::exception& e) {
        json err = {{"type", "error"}, {"code", "BRIDGE_ERROR"}, {"message", e.what()}};
        return err.dump();
    }
}

std::string Bridge::handleCalculate(const std::string& expression, const std::string& angleUnit) {
    m_calc.setAngleUnit(angleUnit == "radians" ? AngleUnit::RADIANS : AngleUnit::DEGREES);
    auto result = m_calc.calculate(expression);

    if (result.error.hasError) {
        json resp = {
            {"type", "error"},
            {"code", result.error.errorCode},
            {"message", result.error.message}
        };
        return resp.dump();
    }

    // Format the value - use integer display when exact
    std::string display;
    double intpart;
    if (std::modf(result.value, &intpart) == 0.0 && std::abs(result.value) < 1e15) {
        std::ostringstream ss;
        ss.precision(15);
        ss << (long long)result.value;
        display = ss.str();
    } else {
        std::ostringstream ss;
        ss.precision(10);
        ss << result.value;
        display = ss.str();
        // Remove trailing zeros
        auto dot = display.find('.');
        if (dot != std::string::npos) {
            auto last = display.find_last_not_of('0');
            if (last > dot) display = display.substr(0, last + 1);
            else display = display.substr(0, dot);
        }
    }

    json resp = {
        {"type", "result"},
        {"value", display},
        {"fullValue", result.value}
    };
    return resp.dump();
}

std::string Bridge::handleMemory(const std::string& operation, double value) {
    if (operation == "store") m_calc.memoryStore(value);
    else if (operation == "add") m_calc.memoryAdd(value);
    else if (operation == "sub") m_calc.memorySub(value);
    else if (operation == "recall") { /* nothing */ }
    else if (operation == "clear") m_calc.memoryClear();

    json resp = {
        {"type", "memory"},
        {"value", m_calc.memoryRecall()},
        {"hasValue", m_calc.hasMemory()}
    };
    return resp.dump();
}

std::string Bridge::handleAngleUnit(const std::string& unit) {
    AngleUnit au = (unit == "radians") ? AngleUnit::RADIANS : AngleUnit::DEGREES;
    m_calc.setAngleUnit(au);
    json resp = {{"type", "angleUnit"}, {"unit", unit}};
    return resp.dump();
}

std::string Bridge::escapeJson(const std::string& s) {
    json j = s;
    return j.dump();
}

File: ui/src/js/bridge.js
// CalcPro — C++ Communication Bridge

export async function calculate(expression, angleUnit) {
  // Try C++ backend first
  if (window.calculate) {
    try {
      const result = await window.calculate(expression, angleUnit || 'degrees');
      return JSON.parse(result);
    } catch {
      // Fall through to JS fallback
    }
  }
  // JS fallback
  const { evaluate } = await import('./calculator.js');
  const res = evaluate(expression, angleUnit || 'degrees');
  if (res.error) {
    return { type: 'error', code: 'EVAL_ERROR', message: res.error };
  }
  return { type: 'result', value: formatValue(res.value), fullValue: res.value };
}

export async function memoryOp(operation, value) {
  if (window.memoryOp) {
    try {
      const result = await window.memoryOp(operation, value);
      return JSON.parse(result);
    } catch {
      // fallback
    }
  }
  return { type: 'memory', value: 0, hasValue: false };
}

export async function setAngleUnit(unit) {
  if (window.setAngleUnit) {
    try {
      await window.setAngleUnit(unit);
    } catch {
      // fallback
    }
  }
}

function formatValue(num) {
  if (Number.isInteger(num) && Math.abs(num) < 1e15) return String(num);
  const s = num.toPrecision(10);
  // Remove trailing zeros
  const dot = s.indexOf('.');
  if (dot !== -1) {
    const trimmed = s.replace(/\.?0+$/, '');
    return trimmed;
  }
  return s;
}
